includes/admin/class-sensei-learners-admin-bulk-actions-controller.php

```diff
@@ -0,0 +1,413 @@
+			$this->redirect_to_learner_admin_index( 'error-invalid-action' );
+			return;
+		}
+
+		foreach ( $course_ids as $course_id ) {
+			// Validate courses before continuing.
+			$course = get_post( absint( $course_id ) );
+			if ( empty( $course ) ) {
+				$this->redirect_to_learner_admin_index( 'error-invalid-course' );
+				return;
+			}
+		}
+
+		foreach ( $user_ids as $user_id ) {
+			$user = new WP_User( $user_id );
+
+			if ( $user->exists() ) {
+				foreach ( $course_ids as $course_id ) {
+					$this->do_user_action( $user_id, (int) $course_id, $sensei_bulk_action );
+				}
+			}
+		}
+
+		$this->redirect_to_learner_admin_index( 'action-success' );
+	}
+
+	/**
+	 * Wraps global method to facilitate mocking.
+	 *
+	 * @access private
+	 */
+	public function check_nonce() {
+		check_admin_referer( self::NONCE_SENSEI_BULK_LEARNER_ACTIONS, self::SENSEI_BULK_LEARNER_ACTIONS_NONCE_FIELD );
+	}
+
+	/**
+	 * Helper method to perform the action on a user and course.
+	 *
+	 * @param integer $user_id   The user to perform the action on.
+	 * @param integer $course_id The course which the action relates to.
+	 * @param string  $action    The action.
+	 */
+	private function do_user_action( $user_id, $course_id, $action ) {
+		try {
+			$manual_enrolment_provider = Sensei_Course_Enrolment_Manager::instance()->get_manual_enrolment_provider();
+		} catch ( Exception $e ) {
+			$this->redirect_to_learner_admin_index( $e->getMessage() );
+		}
+
+		switch ( $action ) {
+			case self::MANUALLY_ENROL:
+				if ( ! $manual_enrolment_provider->is_enrolled( $user_id, $course_id ) ) {
+					$manual_enrolment_provider->enrol_student( $user_id, $course_id );
+				}
+				break;
+			case self::REMOVE_MANUAL_ENROLMENT:
+				if ( $manual_enrolment_provider->is_enrolled( $user_id, $course_id ) ) {
+					$manual_enrolment_provider->withdraw_student( $user_id, $course_id );
+				}
+				break;
+			case self::REMOVE_PROGRESS:
+				if ( Sensei_Utils::has_started_course( $course_id, $user_id ) ) {
+					Sensei_Utils::reset_course_for_user( $course_id, $user_id );
+				}
+				break;
+			case self::COMPLETE_COURSE:
+				if ( Sensei_Utils::has_started_course( $course_id, $user_id ) && ! Sensei_Utils::user_completed_course( $course_id, $user_id ) ) {
+					Sensei_Utils::user_complete_course( $course_id, $user_id );
+				}
+				break;
+			case self::RECALCULATE_COURSE_COMPLETION:
+				if ( Sensei_Utils::has_started_course( $course_id, $user_id ) && ! Sensei_Utils::user_completed_course( $course_id, $user_id ) ) {
+					Sensei_Utils::user_complete_course( $course_id, $user_id, false );
+				}
+				break;
+		}
+	}
+
+	/**
+	 * Enqueues JS and CSS dependencies.
+	 *
+	 * @access private
+	 */
+	public function enqueue_scripts() {
+		$is_debug = defined( 'SCRIPT_DEBUG' ) && SCRIPT_DEBUG;
+
+		wp_enqueue_script(
+			'sensei-admin-jquery-modal',
+			'https://cdnjs.cloudflare.com/ajax/libs/jquery-modal/0.9.1/jquery.modal.min.js',
```

Comment:
Plugin guidelines don't allow us to use 3rd party CDNs for JS https://developer.wordpress.org/plugins/wordpress-org/detailed-plugin-guidelines/#8-plugins-may-not-send-executable-code-via-third-party-systems

I think WPJM does it for JQuery UI CSS, but we should probably not do any more.

Reply:
Wasn't aware of that, I'll readd it! I will probably need to do the same in WCPC since I have also used cdn there. I think that we can use [this one](https://jqueryui.com/dialog/) however it should be probably done as part of another PR. Do you agree?

Reply:
Fixed in c4de3950e516c4287b569465cf7350e2dc54f83b

Reply:
Updated fix which uses `wp_register_script` in 2c6013dcaec439dd06c39a9c076d41434bf935fd